### PR TITLE
Add / Update unit tests

### PR DESCRIPTION
**Pull request by GOVINDDIXIT**
This PR is specific to GSoC'19 project (Improve test automation in Rocket.Chat Android Repo)
- [x] Authentication Module
- [x] Main Module
- [x] SortingandGrouping Module
- [x] chatDetails Module
- [x] chatRooms Module
- [x] createChannel Module
- [x] members Module
- [x] userDetails Module

### CODE REVIEW

File: app/src/main/java/testConfig/Config.kt [hunk 1]
```diff
@@ -36,6 +36,7 @@ class Config {
         const val userName: String = "userName"
         const val avatar: String = "serverUrl/avatar/userName?format=jpeg&rc_uid=null&rc_token=null"
         const val userAvatar: String = "$currentServer/avatar/userName?format=jpeg&rc_uid=null&rc_token=null"
+        const val UPDATED_AVATAR: String = "$currentServer/avatar/$USERNAME?format=jpeg&rc_uid=null&rc_token=null"
```
Comment (ear-dev) [inline app/src/main/java/testConfig/Config.kt]: Govind, small thing.....  this file uses a mix of camelCase and ALL_CAPS.  Is there a reason for that?  Perhaps it should be uniform?

File: app/src/test/java/chat/rocket/android/authentication/loginoptions/presentation/LoginOptionsPresenterTest.kt [hunk 2]
```diff
@@ -52,4 +72,57 @@ class LoginOptionsPresenterTest {
...
+    @Test
+    fun `authenticate user with Saml`() {
+        kotlinx.coroutines.runBlocking {
+            val result = loginOptionsPresenter.authenticateWithSaml(SAML_TOKEN)
+            assertEquals(result, Unit)
+        }
+    }
+
+    @Test
+    fun `authenticate user with Deeplink`() {
```
Comment (ear-dev) [inline app/src/test/java/chat/rocket/android/authentication/loginoptions/presentation/LoginOptionsPresenterTest.kt]: I'm seeing an exception being thrown for all of these coroutine tests, and yet, after the exception the test passes.... do you know what's going on?

```chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenterTest > authenticate user with Deeplink STANDARD_ERROR
    Exception in thread "DefaultDispatcher-worker-7 @coroutine#16" kotlin.UninitializedPropertyAccessException: lateinit property client has not been initialized
        at chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenter.access$getClient$p(LoginOptionsPresenter.kt:44)
        at chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenter$doAuthentication$1$token$1.invokeSuspend(LoginOptionsPresenter.kt:112)
        at chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenter$doAuthentication$1$token$1.invoke(LoginOptionsPresenter.kt)
        at chat.rocket.android.util.IOKt.retryIO(IO.kt:25)
        at chat.rocket.android.util.IOKt.retryIO$default(IO.kt:18)
        at chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenter$doAuthentication$1.invokeSuspend(LoginOptionsPresenter.kt:110)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)

chat.rocket.android.authentication.loginoptions.presentation.LoginOptionsPresenterTest > authenticate user with Deeplink PASSED```

File: app/src/main/java/testConfig/Config.kt (hunk 1 above)
Comment (GOVINDDIXIT) [reply app/src/main/java/testConfig/Config.kt]: Actually, Earlier I was thinking to make some difference in the two types of the variable(Plain string variable with a value similar to name and a variable with some info value). But It is not a good idea I realized. 😜
I will update them.

File: app/src/test/java/chat/rocket/android/authentication/loginoptions/presentation/LoginOptionsPresenterTest.kt (hunk 2 above)
Comment (GOVINDDIXIT) [reply app/src/test/java/chat/rocket/android/authentication/loginoptions/presentation/LoginOptionsPresenterTest.kt]: We still do not have a solution to unit test coroutine methods, Shailesh's solution has solved the null context issue but it is not helpful in testing coroutines methods. I had also discussed with him personally. At the same time, these coroutines methods are very well tested with UI tests so there is no problem. Also, we are now testing the private methods individually using Java reflection which had reduced our dependence on coroutines methods. I have removed the unit tests for coroutines methods as they are not working as expected.

File: app/src/test/java/chat/rocket/android/createchannel/presentation/CreateChannelPresenterTest.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    private val view = mock(CreateChannelView::class.java)
+    private val strategy = mock(CancelStrategy::class.java)
+    private val mapper = mock(MemberUiModelMapper::class.java)
+    private val navigator = mock(MainNavigator::class.java)
+    private val serverInteractor = mock(GetCurrentServerInteractor::class.java)
+    private val factory = mock(RocketChatClientFactory::class.java)
+
+    private lateinit var createChannelPresenter: CreateChannelPresenter
+
```
Comment (ear-dev) [inline app/src/test/java/chat/rocket/android/createchannel/presentation/CreateChannelPresenterTest.kt:24]: extra empty line here

File: app/src/test/java/chat/rocket/android/userdetails/presentation/UserDetailsPresenterTest.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import chat.rocket.android.server.infrastructure.ConnectionManagerFactory
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mockito
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.verify
+import org.mockito.MockitoAnnotations
+import testConfig.Config.Companion.AVATAR_URL
+import testConfig.Config.Companion.CURRENT_SERVER
+
```
Comment (ear-dev) [inline app/src/test/java/chat/rocket/android/userdetails/presentation/UserDetailsPresenterTest.kt:19]: extra empty line here

**Review by ear-dev — APPROVED** [review]
Looks good!  Approved......

**Comment by GOVINDDIXIT** [comment]
Thanks, @ear-dev Will remove the extra lines and merge this one

**Pull request merged**